### DISCUSSION TITLE
Primitive Layer Help

### DISCUSSION THREAD

**Discussion by ilan-gold** [visgl/deck.gl#5739] [Q&A]
Hello,

I am trying to make a simple layer for rendering a cube but nothing appears on my screen (trying to follow the [docs](https://deck.gl/docs/developer-guide/custom-layers/primitive-layers) for this).  I think the issue is that I'm a little unclear on how the `CubeGeometry` attributes can be transferred to the `AttributeManager` - specifically `positions`:

```javascript
 class CubeLayer extends Layer {
  getShaders() {
    return {
      fs,
      vs,
      modules: [project32]
    };
  }

  initializeState() {
    const {gl} = this.context;
    this.setState({model: this._getModel(gl)});
    const attributeManager = this.getAttributeManager();
    const noAlloc = true;
    attributeManager.add({
      positions: {
        size: 3,
        noAlloc
      },
      texCoords: {
        size: 2,
        noAlloc
      }
    });
  }

  _getModel(gl) {
    if (!gl) {
      return null;
    }
    return new Model(gl, {
      ...this.getShaders(),
      id: this.props.id + '-model',
      isInstanced: true,
      geometry: new CubeGeometry()
    });
  }

  draw({uniforms}) {
    const {model} = this.state;
    if (model) {
      model
        .setUniforms(uniforms)
        .draw();
    }
  }
};
```

I found an example of a simple primitive layer:https://github.com/visgl/deck.gl/blob/1049956d723fe7a92c39ea203bc2b5c5fe00cc36/modules/aggregation-layers/src/heatmap-layer/triangle-layer.js that also appears to not need an accessor for `positions` (although the parent layer appears to provide the data).  Do I need to add some sort of `getPosition` function to my layer?  I really don't need explicit positions (those provided by the geometry work I just don't know how to hook them up to the attribute manager).

Thanks!

**Comment by ilan-gold** [comment]
I see that currently the attributes are not binding but by removing 
```javascript
attributeManager.add({
      positions: {
        size: 3,
        noAlloc
      },
      texCoords: {
        size: 2,
        noAlloc
      }
    });
```
they do bind now. Yet, nothing seems to be displaying.

**Reply by ibgreen** [reply]
Just quick thoughts:
- Worth checking that it is not just a size issue - the view and also the coordinate system of the layer - a unit cube could be only 1x1m so would need to zoom in a lot if you are using cartographic views.
- The SimpleMeshLayer and the ScenegraphLayer are set up to draw geometries. Not sure if it would help to see that things show up there.
- Can't see your shaders etc. Perhaps a code pen would help us see what you are doing?

**Reply by ilan-gold** [reply]
@ibgreen I'll try a codepen - in the past I've had trouble with custom layers there but I'll give it a shot again.  I can also try out that other app.....forgetting it's name but it gives a full react component so I'll try that as well.

**Reply by ilan-gold** [reply]
@ibgreen Found the app and made a little demo: https://codesandbox.io/s/cube-layer-example-ijx6h?file=/src/App.js
